src/ray/raylet/scheduling/fair_scheduling_queue.h

```diff
@@ -0,0 +1,118 @@
+#pragma once
+
+/* #include <tuple> */
+#include "ray/common/task/task.h"
+#include "ray/rpc/node_manager/node_manager_server.h"
+
+namespace ray {
+namespace raylet {
+
+/// Work represents all the information needed to make a scheduling decision.
+/// This includes the task, the information we need to communicate to
+/// dispatch/spillback and the callback to trigger it.
+using Work = std::tuple<Task, rpc::RequestWorkerLeaseReply *, std::function<void(void)>>;
+
+namespace internal {
+
+struct SchedulingPriorityComparator : public std::less<SchedulingClass> {
+ public:
+  explicit SchedulingPriorityComparator(
+      std::unordered_map<SchedulingClass, uint64_t> &priorities);
+
+  bool operator()(const SchedulingClass &lhs, const SchedulingClass &rhs) const;
+
+ private:
+  /// A reference to FairSchedulingQueue::active_works_
+  std::unordered_map<SchedulingClass, uint64_t> &priorities_;
+};
+
+using WorkQueueType =
+    std::map<SchedulingClass, std::deque<Work>, internal::SchedulingPriorityComparator>;
+}  // namespace internal
+
+using WorkQueueIterator = internal::WorkQueueType::iterator;
+using ConstWorkQueueIterator = internal::WorkQueueType::const_iterator;
+
+/// A fair scheduling queue. Fairness is defined in terms of the number of
+/// running tasks of a given scheduling class. For example, assume we push 3
+/// tasks with scheduling classes: {1, 2, 2}. Then we mark a task of class 1 as
```

Comment:
I feel a little bit hard to follow here, once 1 is marked as running, there are only 2s left, right? so no matter what we'll schedule class 2 I think?